ddtrace/internal/_queue.pyx

```diff
@@ -0,0 +1,57 @@
+import threading
+
+from ..internal.logger import get_logger
+from . import _rand
+
+
+log = get_logger(__name__)
+
+
+class Empty(Exception):
+    pass
+
+
+class TraceQueue(object):
+    def __init__(self, maxsize=0):
+        self._lock = threading.Lock()
+        self._queue = []
+        self.maxsize = maxsize
+
+        self.accepted = 0
+        self.accepted_lengths = 0
+        self.dropped = 0
+
+    def __len__(self):
+        return len(self._queue)
+
+    def put(self, item):
+        with self._lock:
+            if len(self._queue) < self.maxsize:
+                self._queue.append(item)
+            else:
+                idx = _rand.rand64bits() % len(self._queue)
+                self._queue[idx] = item
+
+                log.warning("Writer queue is full has more than %d traces, dropping trace", self.maxsize)
+                self.dropped += 1
+
+            self.accepted += 1
+            self.accepted_lengths += len(item) if hasattr(item, "__len__") else 1
+
+    def get(self):
+        with self._lock:
+            try:
+                if len(self._queue) == 0:
+                    raise Empty()
+                return self._queue
+            finally:
+                self._queue = []
+
```

Comment:
Not sure why we raise `Empty` here, we could just return an empty list.

Reply:
lol just a relic still of the old/python stdlib implementation - can do away with it again.